language/functional_tests/src/checker/matcher.rs

```diff
@@ -0,0 +1,182 @@
+}
+
+#[derive(Debug)]
+pub enum MatchError {
+    NegativeMatch(Match),
+    UnmatchedDirectives(Vec<usize>),
+    UnmatchedErrors(Vec<usize>),
+}
+
+#[derive(Debug)]
+pub enum MatchStatus {
+    Success,
+    Failure(Vec<MatchError>),
+}
+
+/// The result of matching the directives against the evaluation log.
+#[derive(Debug)]
+pub struct MatchResult {
+    pub status: MatchStatus,
+    pub text: Vec<String>,
+    pub matches: Vec<Match>,
+}
+
+/// Matches the directives against the evaluation log.
+pub fn match_output<I, D>(log: &EvaluationLog, directives: I) -> MatchResult
+where
+    D: AsRef<Directive>,
+    I: IntoIterator<Item = D>,
+{
+    // Convert each entry of the evaluation log into a string, which will be later matched against.
+    let text: Vec<_> = log
+        .outputs
+        .iter()
+        .map(|output| format!("{:?}", output))
+        .collect();
+
+    // Split directives into groups and build an Aho-Corasick automaton for each group.
+    let groups = group_directives(directives);
+    let automata: Vec<_> = groups
+        .iter()
+        .map(|directives| {
+            AhoCorasickBuilder::new()
+                .dfa(true)
+                .build(directives.iter().map(|(_, d)| d.as_ref().pattern_str()))
+        })
+        .collect();
+
+    // Match the groups against the text and record all matches.
+    let mut text_pos = (0, 0);
+    let mut group_pos = 0;
+    let mut matches = vec![];
+    let mut is_log_entry_matched: Vec<_> = text.iter().map(|_| false).collect();
+
+    while text_pos.0 < text.len() && group_pos < groups.len() {
+        match automata[group_pos].earliest_find(&text[text_pos.0][text_pos.1..]) {
+            Some(mat) => {
+                // A match was found.
+                let pat_id = mat.pattern();
+
+                let m = Match {
+                    pat_id: groups[group_pos][pat_id].0,
+                    entry_id: text_pos.0,
+                    start: mat.start() + text_pos.1,
+                    end: mat.end() + text_pos.1,
+                };
+
+                // Abort if the match is negative.
+                if groups[group_pos][pat_id].1.as_ref().is_negative() {
+                    return MatchResult {
+                        status: MatchStatus::Failure(vec![MatchError::NegativeMatch(m)]),
+                        text,
+                        matches,
+                    };
+                }
+
+                // Record the match and move on to the next group of directives.
+                is_log_entry_matched[text_pos.0] = true;
+                matches.push(m);
+                group_pos += 1;
+                text_pos.1 += mat.end();
+                if text_pos.1 >= text[text_pos.0].len() {
+                    text_pos = (text_pos.0 + 1, 0)
+                }
+            }
+            None => {
+                // No match in this entry => move on to the next in the log.
+                text_pos.0 += 1;
+                text_pos.1 = 0;
+            }
```

Comment:
This is A LOT of manual index munging, perhaps you can wrap this in some sort of API to make this cleaner? An iterator of sorts? Just something that controls the advancing? Just a thought

Reply:
Good point. A bit unsure if two-pointer patterns like this would fit nicely into an iterator, but I'll try very hard. This is the piece of code that annoys me myself the most.